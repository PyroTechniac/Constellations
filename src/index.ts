import * as Constants from './lib/util/Constants';
import type { Language } from './lib/structures/Language';
import type { Settings } from './lib/settings/structures/Settings';
import type { GatewayDriver } from './lib/settings/gateway/GatewayDriver';
import type { KeyedObject } from './lib/util/SharedTypes';
import type { SettingsUpdateContext } from './lib/settings/structures/SettingsFolder';

export * from './lib/Client';

// #region extensions

export * from './lib/extensions/ConstellationUser';

// #endregion extensions

// #region settings

export * from './lib/settings/gateway/Gateway';
export * from './lib/settings/gateway/GatewayDriver';
export * from './lib/settings/gateway/GatewayStorage';
export * from './lib/settings/schema/Schema';
export * from './lib/settings/schema/SchemaEntry';
export * from './lib/settings/schema/SchemaFolder';
export * from './lib/settings/structures/Settings';
export * from './lib/settings/structures/SettingsFolder';

// #endregion settings

// #region structures

export * from './lib/structures/base/AliasPiece';
export * from './lib/structures/base/AliasStore';
export * from './lib/structures/base/Piece';
export * from './lib/structures/base/Store';
export * from './lib/structures/Language';
export * from './lib/structures/LanguageStore';
export * from './lib/structures/Provider';
export * from './lib/structures/ProviderStore';
export * from './lib/structures/Serializer';
export * from './lib/structures/SerializerStore';
export * from './lib/structures/SQLProvider';

// #endregion structures

// #region utils

export { Constants };
export * from './lib/util/Colors';
export * from './lib/util/ConstellationConsole';
export * from './lib/util/Cron';
export * from './lib/util/Duration';
export * from './lib/util/QueryBuilder';
export * from './lib/util/RateLimit';
export * from './lib/util/RateLimitManager';
export * from './lib/util/RequestHandler';
export * from './lib/util/SharedTypes';
export * from './lib/util/Stopwatch';
export * from './lib/util/Timestamp';
export * from './lib/util/Type';
export * from './lib/util/Utils';

// #endregion utils

// #region augments

/** @internal */
declare module 'discord.js' {
	interface Client {
		gateways: GatewayDriver;
	}

	interface Guild {
		readonly language: Language;
	}

	interface User {
		settings: Settings;
	}

	interface ClientEvents {
		settingsSync: [Settings];
		settingsDelete: [Settings];
		settingsCreate: [Settings, KeyedObject, SettingsUpdateContext];
		settingsUpdate: [Settings, KeyedObject, SettingsUpdateContext];
	}
}

// #endregion augments