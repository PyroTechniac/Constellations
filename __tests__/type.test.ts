import { Type } from '../src';

const noop = (): void => { }; // eslint-disable-line @typescript-eslint/no-empty-function

describe('array', (): void => {
	test('empty', (): void => {
		const type = new Type([]);
		expect(type.toString()).toBe('Array');
		expect(type).toMatchSnapshot();
	});

	test('same-type', (): void => {
		const type = new Type([1, 2, 3]);
		expect(type.toString()).toBe('Array<number>');
		expect(type).toMatchSnapshot();
	});

	test('different-type', (): void => {
		const type = new Type([true, 'Test', 7])
		expect(type.toString()).toBe('Array<boolean | number | string>');
		expect(type).toMatchSnapshot();
	});

	test('circular', (): void => {
		const a = [[]];
		// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
		// @ts-ignore
		a[0].push(a);
		const type = new Type(a);
		expect(type.toString()).toBe('Array<Array<[Circular:Array]>>');
		expect(type).toMatchSnapshot();
	});
});

describe('function', (): void => {
	test('empty', (): void => {
		const type = new Type(noop)
		expect(type.toString()).toBe('Function(0-arity)');
		expect(type).toMatchSnapshot();
	});

	test('two-args', (): void => {
		// expect(new Type((a: number, b: number): number => a + b).toString()).toBe('Function(2-arity)');
		const type = new Type((a: number, b: number): number => a + b);
		expect(type.toString()).toBe('Function(2-arity)');
		expect(type).toMatchSnapshot();
	});
});

describe('map', (): void => {
	test('empty', (): void => {
		const type = new Type(new Map());
		expect(type.toString()).toBe('Map');
		expect(type).toMatchSnapshot();
	});

	test('same-type', (): void => {
		const type = new Type(new Map([
			['one', 1],
			['two', 2],
			['three', 3]
		]));
		expect(type.toString()).toBe('Map<string, number>');
		expect(type).toMatchSnapshot()
	});

	test('different-type', (): void => {
		const type = new Type(new Map<string, string | number>([
			['text', 'abc'],
			['digit', 123]
		]));
		expect(type.toString()).toBe('Map<string, number | string>');
		expect(type).toMatchSnapshot();
	});

	test('mixed-with-object', (): void => {
		const type = new Type(new Map<string, any>([
			['text', 'abc'],
			['digit', 1],
			['object', {}]
		]));
		expect(type.toString()).toBe('Map<string, Record | number | string>');
		expect(type).toMatchSnapshot();
	});
});

describe('object', (): void => {
	test('generic', (): void => {
		const type = new Type({});
		expect(type.toString()).toBe('Record');
		expect(type).toMatchSnapshot();
	});

	test('null', (): void => {
		const type = new Type(null);
		expect(type.toString()).toBe('null');
		expect(type).toMatchSnapshot();
	});

	test('custom', (): void => {
		class Foo { }
		const type = new Type(new Foo());
		expect(type.toString()).toBe('Foo');
		expect(type).toMatchSnapshot();
	});

	test('types', (): void => {
		const type = new Type({
			foo: 'bar',
			baz: 2,
			hello: true
		});
		expect(type.toString()).toBe('Record<string, boolean | number | string>');
		expect(type).toMatchSnapshot()
	});

	test('recursive', (): void => {
		const type = new Type({
			foo: 'bar',
			hello: {
				baz: 'world'
			}
		});
		expect(type.toString()).toBe('Record<string, Record<string, string> | string>');
		expect(type).toMatchSnapshot();
	});
});

describe('promise', (): void => {
	test('resolves', (): void => {
		const type = new Type(Promise.resolve());

		expect(type.toString()).toBe('Promise<void>');
		expect(type).toMatchSnapshot();
	});
});

describe('set', (): void => {
	test('empty', (): void => {
		const type = new Type(new Set());
		expect(type.toString()).toBe('Set');
		expect(type).toMatchSnapshot();
	});

	test('same-type', (): void => {
		const type = new Type(new Set([1, 2, 3]));
		expect(type.toString()).toBe('Set<number>');
		expect(type).toMatchSnapshot();
	});

	test('different-types', (): void => {
		const type = new Type(new Set(['abc', 123]));
		expect(type.toString()).toBe('Set<number | string>');
		expect(type).toMatchSnapshot();
	});
});
