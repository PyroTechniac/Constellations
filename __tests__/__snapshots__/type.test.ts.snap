// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`array circular 1`] = `
Type {
  "childKeys": Map {},
  "childValues": Map {
    "Array" => Type {
      "childKeys": Map {},
      "childValues": Map {
        "Array" => Type {
          "childKeys": Map {},
          "childValues": Map {},
          "is": "[Circular:Array]",
          "parent": [Circular],
          "value": Array [
            Array [
              [Circular],
            ],
          ],
        },
      },
      "is": "Array",
      "parent": [Circular],
      "value": Array [
        Array [
          [Circular],
        ],
      ],
    },
  },
  "is": "Array",
  "parent": null,
  "value": Array [
    Array [
      [Circular],
    ],
  ],
}
`;

exports[`array different-type 1`] = `
Type {
  "childKeys": Map {},
  "childValues": Map {
    "boolean" => Type {
      "childKeys": Map {},
      "childValues": Map {},
      "is": "boolean",
      "parent": [Circular],
      "value": true,
    },
    "string" => Type {
      "childKeys": Map {},
      "childValues": Map {},
      "is": "string",
      "parent": [Circular],
      "value": "Test",
    },
    "number" => Type {
      "childKeys": Map {},
      "childValues": Map {},
      "is": "number",
      "parent": [Circular],
      "value": 7,
    },
  },
  "is": "Array",
  "parent": null,
  "value": Array [
    true,
    "Test",
    7,
  ],
}
`;

exports[`array empty 1`] = `
Type {
  "childKeys": Map {},
  "childValues": Map {},
  "is": "Array",
  "parent": null,
  "value": Array [],
}
`;

exports[`array same-type 1`] = `
Type {
  "childKeys": Map {},
  "childValues": Map {
    "number" => Type {
      "childKeys": Map {},
      "childValues": Map {},
      "is": "number",
      "parent": [Circular],
      "value": 3,
    },
  },
  "is": "Array",
  "parent": null,
  "value": Array [
    1,
    2,
    3,
  ],
}
`;

exports[`function empty 1`] = `
Type {
  "childKeys": Map {},
  "childValues": Map {},
  "is": "Function(0-arity)",
  "parent": null,
  "value": [Function],
}
`;

exports[`function two-args 1`] = `
Type {
  "childKeys": Map {},
  "childValues": Map {},
  "is": "Function(2-arity)",
  "parent": null,
  "value": [Function],
}
`;

exports[`map different-type 1`] = `
Type {
  "childKeys": Map {
    "string" => Type {
      "childKeys": Map {},
      "childValues": Map {},
      "is": "string",
      "parent": [Circular],
      "value": "digit",
    },
  },
  "childValues": Map {
    "string" => Type {
      "childKeys": Map {},
      "childValues": Map {},
      "is": "string",
      "parent": [Circular],
      "value": "abc",
    },
    "number" => Type {
      "childKeys": Map {},
      "childValues": Map {},
      "is": "number",
      "parent": [Circular],
      "value": 123,
    },
  },
  "is": "Map",
  "parent": null,
  "value": Map {
    "text" => "abc",
    "digit" => 123,
  },
}
`;

exports[`map empty 1`] = `
Type {
  "childKeys": Map {},
  "childValues": Map {},
  "is": "Map",
  "parent": null,
  "value": Map {},
}
`;

exports[`map mixed-with-object 1`] = `
Type {
  "childKeys": Map {
    "string" => Type {
      "childKeys": Map {},
      "childValues": Map {},
      "is": "string",
      "parent": [Circular],
      "value": "object",
    },
  },
  "childValues": Map {
    "string" => Type {
      "childKeys": Map {},
      "childValues": Map {},
      "is": "string",
      "parent": [Circular],
      "value": "abc",
    },
    "number" => Type {
      "childKeys": Map {},
      "childValues": Map {},
      "is": "number",
      "parent": [Circular],
      "value": 1,
    },
    "Object" => Type {
      "childKeys": Map {},
      "childValues": Map {},
      "is": "Record",
      "parent": [Circular],
      "value": Object {},
    },
  },
  "is": "Map",
  "parent": null,
  "value": Map {
    "text" => "abc",
    "digit" => 1,
    "object" => Object {},
  },
}
`;

exports[`map same-type 1`] = `
Type {
  "childKeys": Map {
    "string" => Type {
      "childKeys": Map {},
      "childValues": Map {},
      "is": "string",
      "parent": [Circular],
      "value": "three",
    },
  },
  "childValues": Map {
    "number" => Type {
      "childKeys": Map {},
      "childValues": Map {},
      "is": "number",
      "parent": [Circular],
      "value": 3,
    },
  },
  "is": "Map",
  "parent": null,
  "value": Map {
    "one" => 1,
    "two" => 2,
    "three" => 3,
  },
}
`;

exports[`object custom 1`] = `
Type {
  "childKeys": Map {},
  "childValues": Map {},
  "is": "Foo",
  "parent": null,
  "value": Foo {},
}
`;

exports[`object generic 1`] = `
Type {
  "childKeys": Map {},
  "childValues": Map {},
  "is": "Record",
  "parent": null,
  "value": Object {},
}
`;

exports[`object null 1`] = `
Type {
  "childKeys": Map {},
  "childValues": Map {},
  "is": "null",
  "parent": null,
  "value": null,
}
`;

exports[`object recursive 1`] = `
Type {
  "childKeys": Map {
    "string" => Type {
      "childKeys": Map {},
      "childValues": Map {},
      "is": "string",
      "parent": [Circular],
      "value": "hello",
    },
  },
  "childValues": Map {
    "string" => Type {
      "childKeys": Map {},
      "childValues": Map {},
      "is": "string",
      "parent": [Circular],
      "value": "bar",
    },
    "Object" => Type {
      "childKeys": Map {
        "string" => Type {
          "childKeys": Map {},
          "childValues": Map {},
          "is": "string",
          "parent": [Circular],
          "value": "baz",
        },
      },
      "childValues": Map {
        "string" => Type {
          "childKeys": Map {},
          "childValues": Map {},
          "is": "string",
          "parent": [Circular],
          "value": "world",
        },
      },
      "is": "Record",
      "parent": [Circular],
      "value": Object {
        "baz": "world",
      },
    },
  },
  "is": "Record",
  "parent": null,
  "value": Object {
    "foo": "bar",
    "hello": Object {
      "baz": "world",
    },
  },
}
`;

exports[`object types 1`] = `
Type {
  "childKeys": Map {
    "string" => Type {
      "childKeys": Map {},
      "childValues": Map {},
      "is": "string",
      "parent": [Circular],
      "value": "hello",
    },
  },
  "childValues": Map {
    "string" => Type {
      "childKeys": Map {},
      "childValues": Map {},
      "is": "string",
      "parent": [Circular],
      "value": "bar",
    },
    "number" => Type {
      "childKeys": Map {},
      "childValues": Map {},
      "is": "number",
      "parent": [Circular],
      "value": 2,
    },
    "boolean" => Type {
      "childKeys": Map {},
      "childValues": Map {},
      "is": "boolean",
      "parent": [Circular],
      "value": true,
    },
  },
  "is": "Record",
  "parent": null,
  "value": Object {
    "baz": 2,
    "foo": "bar",
    "hello": true,
  },
}
`;

exports[`promise resolves 1`] = `
Type {
  "childKeys": Map {},
  "childValues": Map {
    "void" => Type {
      "childKeys": Map {},
      "childValues": Map {},
      "is": "void",
      "parent": [Circular],
      "value": undefined,
    },
  },
  "is": "Promise",
  "parent": null,
  "value": Promise {},
}
`;

exports[`set different-types 1`] = `
Type {
  "childKeys": Map {},
  "childValues": Map {
    "string" => Type {
      "childKeys": Map {},
      "childValues": Map {},
      "is": "string",
      "parent": [Circular],
      "value": "abc",
    },
    "number" => Type {
      "childKeys": Map {},
      "childValues": Map {},
      "is": "number",
      "parent": [Circular],
      "value": 123,
    },
  },
  "is": "Set",
  "parent": null,
  "value": Set {
    "abc",
    123,
  },
}
`;

exports[`set empty 1`] = `
Type {
  "childKeys": Map {},
  "childValues": Map {},
  "is": "Set",
  "parent": null,
  "value": Set {},
}
`;

exports[`set same-type 1`] = `
Type {
  "childKeys": Map {},
  "childValues": Map {
    "number" => Type {
      "childKeys": Map {},
      "childValues": Map {},
      "is": "number",
      "parent": [Circular],
      "value": 3,
    },
  },
  "is": "Set",
  "parent": null,
  "value": Set {
    1,
    2,
    3,
  },
}
`;
